### PR TITLE
DateDifference in IR shouldn't use negate

### LINKED ISSUE #355: Fix DateDifference IR to remove negate unary op

### PR DESCRIPTION
**Pull request by lesaltzm**
resolves #355 

Also fixes DateDifference implementation in interpreter to return a number of days.

### CODE REVIEW

**Review by ian-legler — APPROVED** [review]

File: src/libraries/Microsoft.PowerFx.Core/IR/IRTranslator.cs [hunk 1]
```diff
@@ -756,9 +756,14 @@ private static IntermediateNode GetAddBinaryOp(IRTranslatorContext context, Texl
                             // Date + '-DateTime' => in days
                             // Date + '-Date' => in days
 
-                            // Ensure that this is really '-Date' - Binding should always catch this, but let's make sure...
-                            Contracts.Assert(node.Right.AsUnaryOpLit().VerifyValue().Op == UnaryOp.Minus);
-                            return new BinaryOpNode(context.GetIRContext(node), BinaryOpKind.DateDifference, left, right);
+                            // Binding produces this as '-Date'. This should be cleaned up when we switch to a proper sub op.
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Core/IR/IRTranslator.cs:759]: >cleaned up when we switch to a proper sub op.  [](http://example.com/codeflow?start=80&length=46)

Tagging [https://github.com/microsoft/Power-Fx/issues/354](https://github.com/microsoft/Power-Fx/issues/354) here.

File: src/libraries/Microsoft.PowerFx.Core/IR/IRTranslator.cs [hunk 2]
```diff
@@ -756,9 +756,14 @@ private static IntermediateNode GetAddBinaryOp(IRTranslatorContext context, Texl
                             // Date + '-DateTime' => in days
                             // Date + '-Date' => in days
 
-                            // Ensure that this is really '-Date' - Binding should always catch this, but let's make sure...
-                            Contracts.Assert(node.Right.AsUnaryOpLit().VerifyValue().Op == UnaryOp.Minus);
-                            return new BinaryOpNode(context.GetIRContext(node), BinaryOpKind.DateDifference, left, right);
+                            // Binding produces this as '-Date'. This should be cleaned up when we switch to a proper sub op. 
+                            if (right is not UnaryOpNode unaryNode || unaryNode.Op != UnaryOpKind.Negate)
+                            {
+                                throw new NotSupportedException();
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Core/IR/IRTranslator.cs:762]: >throw new NotSupportedException(); [](http://example.com/codeflow?start=32&length=34)

Is there a legal expression that can land here?

Or will binding have prevented it and landing here is a compiler bug.

Comment (lesaltzm) [reply src/libraries/Microsoft.PowerFx.Core/IR/IRTranslator.cs:762]: Binding will have prevented this. It's always a bug if we land here.

Comment (lesaltzm) [reply src/libraries/Microsoft.PowerFx.Core/IR/IRTranslator.cs:762]: see 
image

**Pull request merged**